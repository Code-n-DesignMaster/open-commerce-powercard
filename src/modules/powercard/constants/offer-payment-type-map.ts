import { OFFER_PAYMENT_TYPE } from '@open-commerce/data-objects';
import { mapValues, invert } from 'lodash';

// This one-liner generates the following map:
// [
//   OFFER_PAYMENT_TYPE.NONE,
//   OFFER_PAYMENT_TYPE.APPLE_PAY,
//   OFFER_PAYMENT_TYPE.GOOGLE_PAY,
//   ...
// ];
// This code does not need to be modified as new enum values are added.
export const offerPaymentTypeMap = Object.keys(OFFER_PAYMENT_TYPE);

// These lodash functions generate the following map:
// {
//   NONE: 0,
//   APPLE_PAY: 1,
//   GOOGLE_PAY: 2,
//   ...
// };
// This code does not need to be modified as new enum values are added.
export const offerPaymentTypeToNumberMap = mapValues(
  invert(offerPaymentTypeMap),
  v => +v,
);
