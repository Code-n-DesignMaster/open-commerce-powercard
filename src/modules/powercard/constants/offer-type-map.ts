import { OFFER_TYPE } from '@open-commerce/data-objects';
import { mapValues, invert } from 'lodash';

// [
//   OFFER_TYPE.EMAIL,
//   OFFER_TYPE.NEW_USER,
//   OFFER_TYPE.EXISTING_USER,
//   OFFER_TYPE.STORE_ID,
//   OFFER_TYPE.STATE,
//   OFFER_TYPE.CHIP_BALANCE,
//   OFFER_TYPE.TICKET_BALANCE,
//   OFFER_TYPE.APPLE_PAY,
//   OFFER_TYPE.GOOGLE_PAY,
//   ...
// ];
export const offerTypeMap = Object.keys(OFFER_TYPE).filter(
  key => key !== 'NONE',
);

// {
//   EMAIL: 0,
//   NEW_USER: 1,
//   EXISTING_USER: 2,
//   STORE_ID: 3,
//   STATE: 4,
//   CHIP_BALANCE: 5,
//   TICKET_BALANCE: 6,
//   APPLE_PAY: 7,
//   GOOGLE_PAY: 8,
//   ...
// };
export const offerTypeToNumberMap = mapValues(invert(offerTypeMap), v => +v);

// 0	0x0001	1	Email
// 1	0x0002	2	New User
// 2	0x0004	4	Existing User
// 3	0x0008	8	Store ID
// 4	0x0010	16	State
// 5	0x0020	32	Chip Balance
// 6	0x0040	64	Ticket Balance
// 7	0x0080	128	Apple Pay
// 8	0x0100	256	Google Pay
// ...

// {
//   EMAIL: 1,
//   NEW_USER: 2,
//   EXISTING_USER: 4,
//   STORE_ID: 8,
//   STATE: 16,
//   CHIP_BALANCE: 32,
//   TICKET_BALANCE: 64,
//   APPLE_PAY: 128,
//   GOOGLE_PAY: 256,
//   ...
// }
export const offerTypeToBitmaskMap = mapValues(
  offerTypeToNumberMap,
  (value: number) => 2 ** value,
);
