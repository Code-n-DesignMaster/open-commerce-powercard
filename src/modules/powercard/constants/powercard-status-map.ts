import { POWERCARD_STATUS_TYPE } from '@open-commerce/data-objects';

export const powercardStatusMap = [
  null,
  POWERCARD_STATUS_TYPE.DISABLED,
  POWERCARD_STATUS_TYPE.INACTIVE,
  POWERCARD_STATUS_TYPE.OPEN,
  POWERCARD_STATUS_TYPE.STOLEN,
  POWERCARD_STATUS_TYPE.VIP,
  POWERCARD_STATUS_TYPE.VOID,
];
