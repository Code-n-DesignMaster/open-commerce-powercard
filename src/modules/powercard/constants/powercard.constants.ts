export const DEFAULT_TIMEZONE = 'America/Chicago';
export const POWERCARD_REPOSITORY_TOKEN = 'POWERCARD_REPOSITORY_TOKEN';
export const RATING_REPOSITORY_TOKEN = 'RATING_REPOSITORY_TOKEN';
export const POWERCARD_IMAGE_PACK_REPOSITORY_TOKEN =
  'POWERCARD_IMAGE_PACK_REPOSITORY_TOKEN';
export const DB_CONNECTION_TOKEN = 'DB_CONNECTION_TOKEN';
export const TEST_DB_CONNECTION_TOKEN = 'TEST_DB_CONNECTION_TOKEN';
export const TEST_POWERCARD_REPOSITORY_TOKEN =
  'TEST_POWERCARD_REPOSITORY_TOKEN';
export const TRANSACTION_SERVICE_TOKEN = 'TRANSACTION_SERVICE_TOKEN';
export const CUSTOMER_SERVICE_TOKEN = 'CUSTOMER_SERVICE_TOKEN';
export const MOBILE_PASS_SERVICE_TOKEN = 'MOBILE_PASS_SERVICE_TOKEN';
export const POWERCARD_BALANCE_SNAPSHOT_REPOSITORY_TOKEN =
  'POWERCARD_BALANCE_SNAPSHOT_REPOSITORY_TOKEN';
export const TRANSACTION_REPOSITORY_TOKEN = 'TRANSACTION_REPOSITORY_TOKEN';
export const OFFER_REDEMPTION_REPOSITORY_TOKEN =
  'OFFER_REDEMPTION_REPOSITORY_TOKEN';
export const RECEIPT_REPOSITORY_TOKEN = 'RECEIPT_REPOSITORY_TOKEN';
export const NOTIFICATION_SERVICE_CLIENT_TOKEN =
  'NOTIFICATION_SERVICE_CLIENT_TOKEN';
export const TABLE_GUID_REPOSITORY_TOKEN = 'TABLE_GUID_REPOSITORY_TOKEN';
export const RABBITMQ_SERVICE_TOKEN = 'RABBITMQ_SERVICE_TOKEN';
export const RECEIPT_EMAIL_BUILDER_TOKEN = 'RECEIPT_EMAIL_BUILDER_TOKEN';
export const PROMO_IMAGE_REPOSITORY_TOKEN = 'PROMO_IMAGE_REPOSITORY_TOKEN';
