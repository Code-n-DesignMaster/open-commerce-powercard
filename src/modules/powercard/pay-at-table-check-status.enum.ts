export enum PAY_AT_TABLE_CHECK_STATUS {
  OPEN = 1,
  CLOSED = 3,
}
