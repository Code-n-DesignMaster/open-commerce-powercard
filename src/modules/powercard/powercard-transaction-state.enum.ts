export enum POWERCARD_TRANSACTION_STATE {
  AUTHORIZED = 'AUTHORIZED',
  ACTIVATING_VIRTUAL_POWERCARD = 'ACTIVATING_VIRTUAL_POWERCARD',
  CREATING_VIRTUAL_POWERCARD = 'CREATING_VIRTUAL_POWERCARDS',
  UPDATING_CUSTOMER = 'UPDATING_CUSTOMER',
  RECHARGING_POWERCARD = 'RECHARGING_POWERCARD',
  REDEEMING_OFFER = 'REDEEMING_OFFER',
  CREATING_POWERCARD_SNAPSHOT = 'CREATING_POWERCARD_SNAPSHOT',
  VOIDING_TRANSACTION = 'VOIDING_TRANSACTION',
}
