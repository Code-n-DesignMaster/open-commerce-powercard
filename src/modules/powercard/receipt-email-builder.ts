import * as fs from 'fs';
import {
  EmailInputDto,
  Receipt,
  ReceiptLineItem,
  CHECK_LINE_ITEM_TYPE,
  Transaction,
} from '@open-commerce/data-objects';
import * as moment from 'moment-timezone';
import { get } from 'lodash';
import { POWERCARD_SERVICE_CONFIG } from '../../config/config.constants';
import { Inject } from '@nestjs/common';
import { IPowercardServiceConfig } from '../../config/powercard-service-config.interface';

const NUMBER_FORMAT = 'en-US';
const CURRENCY_FORMAT = 'USD';

const RECEIPT_EMAIL_SUBJECT = `Your Dave & Buster's Receipt`;

const PRICE_SUMMARY = [
  {
    description: 'subtotal',
    amount: 'subTotal',
  },
  {
    description: 'tax',
    amount: 'tax',
  },
];

const LINE_ITEMS_SLUG = '~~LINE_ITEMS~~';
const CHECK_NUMBER = '~~CHECK_NUMBER~~';
const PRICE_SUMMARY_SLUG = '~~PRICE_SUMMARY~~';
const TOTAL_PRICE_SLUG = `~~TOTAL_PRICE~~`;
const DATE_SLUG = `~~DATE~~`;
const TENDER_SLUG = `~~TENDER~~`;
const TIP_SLUG = `~~TIP~~`;
const PAYMENT_PROVIDER_TRANSACTION_ID: string =
  '~~PAYMENT_PROVIDER_TRANSACTION_ID~~';

export class ReceiptEmailBuilder {
  private emailFromAddress: string = null;

  public constructor(
    @Inject(POWERCARD_SERVICE_CONFIG)
    private readonly config: IPowercardServiceConfig,
  ) {
    this.initialize();
  }

  public initialize() {
    this.emailFromAddress = this.config.emailFromAddress;

    if (!this.emailFromAddress) {
      throw new Error(
        `You must specify the environment variables OC__EMAIL_FROM_ADDRESS!
                          examples:
                          export OC__EMAIL_FROM_ADDRESS=scott.wasserman@stuzo.com
                          `,
      );
    }
  }

  public async buildEmailInputDto(
    toAddress: string,
    receipt: Receipt,
    timezone: string,
    transaction: Transaction,
  ): Promise<EmailInputDto> {
    const emailInput = new EmailInputDto();
    emailInput.fromAddress = this.emailFromAddress;
    emailInput.toAddress = toAddress;
    emailInput.html = await this.buildReceiptEmailHtml(
      receipt,
      timezone,
      transaction,
    );
    emailInput.subject = RECEIPT_EMAIL_SUBJECT;
    // TODO: add text email?
    return emailInput;
  }

  public fillLineItem(
    description: string,
    amount: number,
    quantity: string | number = '',
  ): string {
    return `
        <tr style="font-size: 12px;border: none;">
            <td style="padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;">
                ${quantity} ${description}
            </td>
            <td style="padding-top: 5px;font-size: 12px; text-align: right;border: none;">
                ${new Intl.NumberFormat(NUMBER_FORMAT, {
                  style: 'currency',
                  currency: CURRENCY_FORMAT,
                }).format(amount)}
            </td>
      </tr>`;
  }

  public getDateHtml(receipt: Receipt, timezone: string): string {
    const { checkOpenTime } = receipt;

    const time = moment(checkOpenTime).tz(timezone);

    return `
        <td style="border: none; text-align: right; font-size: 0.5em;">
          ${time.format('MMMM DD, YYYY')}
        <br>
          ${time.format('hh:mm A')}
        </td>
      `;
  }

  public getLineItemsHtml(lineItems: ReceiptLineItem[]): string {
    let receiptLineItemsHtml: string = '';

    const orderedLineItems: ReceiptLineItem[] = lineItems.filter(item =>
      [CHECK_LINE_ITEM_TYPE.MENU_ITEM, CHECK_LINE_ITEM_TYPE.DISCOUNT].includes(
        item.itemType as CHECK_LINE_ITEM_TYPE,
      ),
    );

    orderedLineItems.forEach(item => {
      const html = this.fillLineItem(
        item.description,
        item.amount,
        item.quantity,
      );

      receiptLineItemsHtml = receiptLineItemsHtml.concat(html);
    });

    return receiptLineItemsHtml;
  }

  public async getTendersHtml(
    receipt: Receipt,
    lineItems: ReceiptLineItem[],
    transaction: Transaction,
  ): Promise<string> {
    let receiptTenderHtml: string = '';
    const tenders: ReceiptLineItem[] = lineItems.filter(
      item => item.itemType === CHECK_LINE_ITEM_TYPE.TENDER,
    );

    for (const tender of tenders) {
      const { amount } = tender;
      let { description } = tender;

      if (description.toLowerCase() === 'unknown') {
        const paymentMethodName = get(
          transaction,
          'paymentInfo.data.0.cardType',
        ) as string;
        description = paymentMethodName;
      }
      const html = this.fillLineItem(description, amount);

      receiptTenderHtml = receiptTenderHtml.concat(html);
    }

    return receiptTenderHtml;
  }

  public getPriceSummaryHtml(receipt: Receipt): string {
    let receiptPriceSummaryHtml = '';

    Object.keys(PRICE_SUMMARY).forEach(key => {
      const description: string = PRICE_SUMMARY[key].description;
      const amount: string = PRICE_SUMMARY[key].amount;
      const html = this.fillLineItem(description, receipt[amount]);

      receiptPriceSummaryHtml = receiptPriceSummaryHtml.concat(html);
    });

    return receiptPriceSummaryHtml;
  }

  public getTipHtml(lineItems: ReceiptLineItem[]): string {
    const tip: ReceiptLineItem = lineItems.find(
      item => item.itemType === CHECK_LINE_ITEM_TYPE.TIP,
    );

    const tipValue = tip ? tip.amount : 0.0;

    return this.fillLineItem('$ CHARGED TIP', tipValue);
  }

  public async buildReceiptEmailHtml(
    receipt: Receipt,
    timezone: string,
    transaction: Transaction,
  ): Promise<string> {
    const { lineItems } = receipt;
    const formattedTotalPrice: string = new Intl.NumberFormat(NUMBER_FORMAT, {
      style: 'currency',
      currency: CURRENCY_FORMAT,
    }).format(receipt.total);

    let emailHtml = fs
      .readFileSync(__dirname + '/receipt_email.html')
      .toString();

    emailHtml = emailHtml.replace(
      new RegExp(CHECK_NUMBER, 'g'),
      receipt.checkNumber + '',
    );

    emailHtml = emailHtml.replace(
      new RegExp(PAYMENT_PROVIDER_TRANSACTION_ID, 'g'),
      transaction.paymentProviderTransactionId + '',
    );

    emailHtml = emailHtml.replace(
      new RegExp(LINE_ITEMS_SLUG, 'g'),
      this.getLineItemsHtml(lineItems),
    );

    emailHtml = emailHtml.replace(
      new RegExp(PRICE_SUMMARY_SLUG, 'g'),
      this.getPriceSummaryHtml(receipt),
    );

    emailHtml = emailHtml.replace(
      new RegExp(TOTAL_PRICE_SLUG, 'g'),
      formattedTotalPrice,
    );

    emailHtml = emailHtml.replace(
      new RegExp(DATE_SLUG, 'g'),
      this.getDateHtml(receipt, timezone),
    );

    emailHtml = emailHtml.replace(
      new RegExp(TIP_SLUG, 'g'),
      this.getTipHtml(lineItems),
    );

    emailHtml = emailHtml.replace(
      new RegExp(TENDER_SLUG, 'g'),
      await this.getTendersHtml(receipt, lineItems, transaction),
    );

    return emailHtml;
  }
}
