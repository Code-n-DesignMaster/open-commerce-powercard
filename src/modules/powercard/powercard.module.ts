import { Module } from '@nestjs/common';
import { MarsCachingModule } from '../mars-caching/mars-caching.module';
import { PowercardService } from './powercard.service';
import { powercardProviders } from './powercard.providers';
import { LoggerModule } from '@open-commerce/nestjs-logger';
import { MarsModule } from '../mars/mars.module';
import {
  OfferRedemption,
  Powercard,
  PowercardImagePack,
  PowercardBalanceSnapshot,
  Transaction,
  Rating,
  Receipt,
  TableGuid,
  PromoImage,
} from '@open-commerce/data-objects';
import { TypeOrmModule } from '@nestjs/typeorm';
import { RabbitmqModule } from '@open-commerce/nestjs-rabbitmq';
import { config } from '../../config/config';
import { ConfigModule } from '@open-commerce/nestjs-config';

@Module({
  imports: [
    ConfigModule.forRoot(config),
    TypeOrmModule.forFeature([
      Powercard,
      PowercardImagePack,
      PowercardBalanceSnapshot,
      Transaction,
      OfferRedemption,
      Rating,
      Receipt,
      PromoImage,
      TableGuid,
    ]),
    RabbitmqModule,
    LoggerModule,
    MarsModule,
    MarsCachingModule,
  ],
  providers: powercardProviders,
  exports: [PowercardService],
})
export class PowercardModule {}
