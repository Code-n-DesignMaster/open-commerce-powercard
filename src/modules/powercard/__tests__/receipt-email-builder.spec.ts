import { TRANSACTION_REPOSITORY_TOKEN } from '../constants/powercard.constants';
import { Test } from '@nestjs/testing';
import { mockReceipt } from './__mocks__/mock-receipt';
import { ReceiptEmailBuilder } from '../receipt-email-builder';
import {
  ReceiptLineItem,
  CHECK_LINE_ITEM_TYPE,
  TRANSACTION_STATUS,
  Transaction,
} from '@open-commerce/data-objects';
import { POWERCARD_SERVICE_CONFIG } from '../../../config/config.constants';

describe('ReceiptEmailBuilder', () => {
  let receiptEmailBuilder: ReceiptEmailBuilder;
  let generatedEmail: string;

  const currency = 'USD';
  const locale = 'en-US';

  const mockTender: ReceiptLineItem = mockReceipt.lineItems.find(
    item => item.itemType === CHECK_LINE_ITEM_TYPE.TENDER,
  );
  const mockTip: ReceiptLineItem = mockReceipt.lineItems.find(
    item => item.itemType === CHECK_LINE_ITEM_TYPE.TIP,
  );
  const mockMenuItem: ReceiptLineItem = mockReceipt.lineItems.find(
    item => item.itemType === CHECK_LINE_ITEM_TYPE.MENU_ITEM,
  );
  const mockReference: ReceiptLineItem = mockReceipt.lineItems.find(
    item => item.itemType === CHECK_LINE_ITEM_TYPE.REFERENCE,
  );

  const mockConfig = {
    powercard: {
      rateCardVersion: -1,
      enableStoreIdOverride: false,
      storeIdOverrideValue: '',
      dontEnablePayAtTableListener: true,
      marsCreditLimitBypass: false,
      emailFromAddress: 'test@email.com',
    },
  };

  const mockTransaction = ({
    id: 1,
    customerUuid: 'test-customer-uuid',
    brandSpecificLocationId: 123,
    tip: 10,
    uuid: 'test-fake-transaction-uuid-authorization-code',
    powercard: {
      cardNumber: '1234',
    },
    paymentProviderTransactionId: 'braintreeTxId',
    paymentInfo: {
      data: [
        {
          cardType: 'mastercard',
        },
      ],
    },
    status: TRANSACTION_STATUS.CLOSED,
  } as unknown) as Transaction; // force casting to Transcation in order not to mock too much

  const mockTransactionRepository = {
    findOneOrFail: jest.fn().mockReturnValueOnce(mockTransaction),
  };

  const mockTimezone = 'America/Chicago';

  beforeEach(async () => {
    const module = await Test.createTestingModule({
      providers: [
        ReceiptEmailBuilder,
        {
          provide: POWERCARD_SERVICE_CONFIG,
          useValue: mockConfig.powercard,
        },
        {
          provide: TRANSACTION_REPOSITORY_TOKEN,
          useValue: mockTransactionRepository,
        },
      ],
    }).compile();

    receiptEmailBuilder = module.get<ReceiptEmailBuilder>(ReceiptEmailBuilder);

    generatedEmail = await receiptEmailBuilder.buildReceiptEmailHtml(
      mockReceipt,
      mockTimezone,
      mockTransaction,
    );
  });

  describe('receipt email builder', () => {
    test('generated email should match to snapshot', async () => {
      expect(generatedEmail).toMatchSnapshot();
    });

    it(`should show receipt's total price`, async () => {
      const totalPrice = new Intl.NumberFormat(locale, {
        style: 'currency',
        currency,
      }).format(mockReceipt.total);

      expect(generatedEmail).toEqual(expect.stringContaining(totalPrice));
      expect(generatedEmail).toEqual(expect.stringContaining('total'));
    });

    it(`should show receipt's menu item `, async () => {
      const menuItemPrice = new Intl.NumberFormat(locale, {
        style: 'currency',
        currency,
      }).format(mockMenuItem.amount);

      expect(generatedEmail).toEqual(
        expect.stringContaining(mockMenuItem.description),
      );
      expect(generatedEmail).toEqual(expect.stringContaining(menuItemPrice));
      expect(generatedEmail).toEqual(
        expect.stringContaining(mockMenuItem.quantity.toString()),
      );
    });
    it(`should show receipt's tip`, async () => {
      const tipPrice = new Intl.NumberFormat(locale, {
        style: 'currency',
        currency,
      }).format(mockTip.amount);

      expect(generatedEmail).toEqual(expect.stringContaining('$ CHARGED TIP'));
      expect(generatedEmail).toEqual(expect.stringContaining(tipPrice));
    });
    it(`should show receipt's tender`, async () => {
      const tenderPrice = new Intl.NumberFormat(locale, {
        style: 'currency',
        currency,
      }).format(mockTender.amount);

      expect(generatedEmail).toEqual(
        expect.stringContaining(mockTender.description),
      );
      expect(generatedEmail).toEqual(expect.stringContaining(tenderPrice));
    });
    it(`should show receipt's actual if tender is unknown`, async () => {
      const tenderPrice = new Intl.NumberFormat(locale, {
        style: 'currency',
        currency,
      }).format(mockTender.amount);

      expect(generatedEmail).toEqual(expect.stringContaining(tenderPrice));
      expect(generatedEmail).toEqual(expect.stringContaining(tenderPrice));
    });
    it(`should not show receipt's reference`, async () => {
      expect(generatedEmail).toEqual(
        // @ts-ignore: Property does not exist
        expect.not.stringMatching(mockReference.reference),
      );
    });
  });
});
