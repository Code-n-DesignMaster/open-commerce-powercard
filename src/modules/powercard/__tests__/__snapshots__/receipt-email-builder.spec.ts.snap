// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReceiptEmailBuilder receipt email builder generated email should match to snapshot 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <meta charset=\\"UTF-8\\" />
    <title>Your Dave & Buster's Receipt</title>

    <style>
      table,
      tr,
      td {
        border: none;
      }
    </style>
  </head>

  <body style=\\"font-family: sans-serif; font-size: 25px\\">
    <table
      style=\\"max-width: 400px; width: 100%;\\"
      align=\\"center\\"
      border=\\"1\\"
      cellpadding=\\"0\\"
      cellspacing=\\"0\\"
      width=\\"300\\"
    >
      <tr>
        <td>
          <img
            src=\\"https://s3.amazonaws.com/dave-busters-public/email-verification/logo.png\\"
            style=\\"height: 50px; width: 50px\\"
          />
        </td>

        
        <td style=\\"border: none; text-align: right; font-size: 0.5em;\\">
          December 01, 2020
        <br>
          04:24 AM
        </td>
      
      </tr>

      <tr style=\\"border: none;\\">
        <td style=\\"padding-top: 10px; font-size: 12px;text-transform: uppercase; color: #5f6368; border: none;\\">
          chk <span style=\\"font-weight: 400;\\">56</span><br />
          txid <span style=\\"font-weight: 400;\\">braintreeTxId</span><br />
        </td>
      </tr>

      <tr style=\\"border: none;\\">
        <td style=\\"font-weight:bold; padding-top: 10px; border: none;\\">
          Total
        </td>
        <td style=\\"padding-top: 10px; text-align: right;border: none;\\">
          $100.00
        </td>
      </tr>

      <tr style=\\"border: none;\\">
        <td colspan=\\"2\\" style=\\"border: none;\\">
          <hr
            style=\\"width: 100%; border-top: 1px solid #EFEFEF; margin-top: 10px; margin-bottom: 10px;\\"
          />
        </td>
      </tr>

      
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                1 DR PEPPER
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $2.99
            </td>
      </tr>

      <tr style=\\"border: none;\\">
        <td colspan=\\"2\\" style=\\"border: none;\\">
          <hr
            style=\\"width: 100%; border-top: 1px solid #EFEFEF; margin-top: 10px; margin-bottom: 10px;\\"
          />
        </td>
      </tr>

      <tr>
        <td colspan=\\"2\\" style=\\"border: none;\\">
          
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                 subtotal
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $75.65
            </td>
      </tr>
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                 tax
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $3.10
            </td>
      </tr>
        </td>
      </tr>

      <tr>
        <td colspan=\\"2\\" style=\\"border: none;\\">
          
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                 $ CHARGED TIP
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $5.00
            </td>
      </tr>
        </td>
      </tr>

      <tr style=\\"border: none;\\">
        <td style=\\"font-weight:bold; padding: 10px 0; border: none;\\">
          Total
        </td>
        <td style=\\"padding: 10px 0; text-align: right;border: none;\\">
          $100.00
        </td>
      </tr>

      
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                 mastercard
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $2.00
            </td>
      </tr>
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                 visa
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $2.00
            </td>
      </tr>
        <tr style=\\"font-size: 12px;border: none;\\">
            <td style=\\"padding-top: 5px;font-size: 12px;text-transform: uppercase;font-weight: bold;border: none;\\">
                 mastercard
            </td>
            <td style=\\"padding-top: 5px;font-size: 12px; text-align: right;border: none;\\">
                $2.00
            </td>
      </tr>
    </table>
  </body>
</html>
"
`;
